### PR TITLE
clean up various files mostly in Widgets\DataTypes

### PR DESCRIPTION
**Pull request by Morbidous**
add a couple extension methods

### CODE REVIEW

File: DownUnder.UI/UI/Widgets/DataTypes/SerializableType.cs [hunk 1]
```diff
@@ -1,114 +1,60 @@
...
 
-        // overload the .Equals method
-        public override bool Equals(System.Object obj)
-        {
-            // If parameter is null return false.
-            if (obj == null)
-            {
-                return false;
-            }
+        // we don't need to overload operators between SerializableType and System.Type because we already enabled them to implicitly convert
```
Comment (jamieyello) [inline DownUnder.UI/UI/Widgets/DataTypes/SerializableType.cs:46]: I take it as a good sign that my most criticized code yet is a copy paste from stackoverflow, tons of good changes

File: DownUnder.UI/UI/Widgets/DataTypes/UIInputState.cs [hunk 2]
```diff
@@ -1,75 +1,74 @@
...
-        private bool _previous_insert_key_down = false;
+    public sealed class UIInputState {
+        Vector2 _previous_mouse_wheel_scroll = new Vector2(Mouse.GetState().HorizontalScrollWheelValue, Mouse.GetState().ScrollWheelValue);
+        MouseState _mouse_state = Mouse.GetState();
+        MouseState _previous_mouse_state = Mouse.GetState();
+        KeyboardState _keyboard_state = Keyboard.GetState();
+        KeyboardState _previous_keyboard_state = Keyboard.GetState();
+        readonly BufferedKeyboard buffered_keyboard = new BufferedKeyboard();
+        bool _insert_key_down;
+        bool _previous_insert_key_down;
```
Comment (jamieyello) [inline DownUnder.UI/UI/Widgets/DataTypes/UIInputState.cs:19]: Private is pretty useless outside of class/struct use despite being one of the the first things we're taught

**Comment by jamieyello** [comment]
Wow, I have no idea how you manage to do so much. Is it more code than you expected?

I guess I've never spent an entire session cleaning code

**Comment by Morbidous** [comment]
I'm on vacation for one thing. 8)

At work I have around 175-200 C# projects in my main solution to maintain. I try to keep them up to date but there's only so much you can do so fast in an 8 hour work day (and programming isn't the only thing I do). What happens is as I work on one system, I learn new things, and decide on new patterns or stop using old patterns, etc., but then only that system gets updated with those new ideas, and everything else that I'm not working on falls way behind. Some of the systems that are still in use today are upgrades of systems I wrote 15-20 years ago.

Also, I am using refactoring tools as well, with all of my code style rules warning me about things that are not the way I want them, many with quick fixes available. Here's a shot of the IDE (JetBrains Rider) if I can figure out how to post an image:

image

(ok I can just paste raw image data right into the post and it uploads and links it automatically, that's cool)

You can see the dark method names meaning that those methods are never called. The white variable names are variables that are written to more than once, which is something to try to avoid (would be nice if C# had readonly variables).

The underlines on the type names are recommendations to use var (based on my settings). The dots under the if are a recommendation to invert the if to reduce nesting (and there is a refactoring command to automatically invert it) (not always a good idea, so you can't just blindly trust the tools).

You can do things like highlight one instance of an expression that you see repeated, and then use the Introduce Variable refactoring and it will grab all occurrences of that expression and replace it with a new variable.

I have actually been suffering with VS + Resharper for the last many years, and it was at times so unbearably slow, but Rider in comparison is lightning fast. It's basically the best IDE I've ever used and I have been *way* more productive since switching to it from VS (it was either around Oct 2020 or 2019 that I switched but I can't remember). (Though, I still need VS for the odd thing here and there, like the package manager console or the RDLC editor.)

**Pull request merged**